service/src/main/java/org/cbioportal/service/impl/GenePanelServiceImpl.java

```diff
@@ -76,118 +80,177 @@ public GenePanel getGenePanel(String genePanelId) throws GenePanelNotFoundExcept
+            }
+        } else {
+            // TODO: remove this block after fusion are migrated to structural variant in database
 
-        // TODO: remove this block after fusion are migrated to structural variant in database
-        List<String> updatedMolecularProfileIds = new ArrayList<>(molecularProfileIds);
-        List<String> updatedSampleIds = new ArrayList<>(sampleIds);
+            List<String> updatedMolecularProfileIds = new ArrayList<>(molecularProfileIds);
+            List<String> updatedSampleIds = new ArrayList<>(sampleIds);
 
-        for (int i = 0; i < molecularProfileIds.size(); i++) {
-            String molecularProfileId = molecularProfileIds.get(i);
-            if(molecularProfileId.endsWith("_fusion")) {
-                updatedMolecularProfileIds.add(molecularProfileId.replace("_fusion", "_mutations"));
-                updatedSampleIds.add(sampleIds.get(i));
+            for (int i = 0; i < molecularProfileIds.size(); i++) {
+                String molecularProfileId = molecularProfileIds.get(i);
+                if(molecularProfileId.endsWith("_fusion")) {
+                    updatedMolecularProfileIds.add(molecularProfileId.replace("_fusion", "_mutations"));
+                    updatedSampleIds.add(sampleIds.get(i));
+                }
             }
-        }
 
-        List<GenePanelData> genePanelData = genePanelRepository
-            .fetchGenePanelDataInMultipleMolecularProfiles(updatedMolecularProfileIds, updatedSampleIds);
+            if(updatedSampleIds.size() < maxCasesCountToIncludeInQuery) {
+                genePanelDataForQueriedProfiles = genePanelRepository
+                    .fetchGenePanelDataInMultipleMolecularProfiles(updatedMolecularProfileIds, updatedSampleIds);
+            } else {
+                distinctMolecularProfileIds = updatedMolecularProfileIds
+                    .stream()
+                    .distinct()
+                    .collect(Collectors.toList());
+                genePanelDataForQueriedProfiles = genePanelRepository.fetchGenePanelDataInMultipleMolecularProfiles(distinctMolecularProfileIds, null);
+            }
+        }
 
-        Map<Pair<String, String>, GenePanelData> genePanelDataSet = genePanelData
+        Map<Pair<String, String>, GenePanelData> genePanelDataSet = genePanelDataForQueriedProfiles
             .stream()
             .collect(toMap(
-                    datum -> new Pair(datum.getMolecularProfileId(),
+                datum -> new Pair(datum.getMolecularProfileId(),
                     datum.getSampleId()), Function.identity()));
 
-        return genePanelData.stream().map(datum -> {
-            String molecularProfileId = datum.getMolecularProfileId();
+        for (int i = 0; i < molecularProfileIds.size(); i++) {
+            String molecularProfileId = molecularProfileIds.get(i);
+            String sampleId = sampleIds.get(i);
```

Comment:
Similar question here - is sampleIds the same size as molecularProfileIds?

Reply:
yes